server/src/main/kotlin/org/javacs/kt/SourcePath.kt

```diff
@@ -148,14 +155,21 @@ class SourcePath(
         // Compile changed files
         fun compileAndUpdate(changed: List<SourceFile>, kind: CompilationKind): BindingContext? {
             if (changed.isEmpty()) return null
-            val parse = changed.map { it.parseIfChanged().parsed!! }
-            val (context, container) = cp.compiler.compileFiles(parse, all(), kind)
+            val parse = changed.associateWith { it.parseIfChanged().parsed!! }
+            val all = all()
+            beforeCompileCallback.invoke()
+            val (context, container) = cp.compiler.compileFiles(parse.values, all, kind)
 
             // Update cache
-            for (f in changed) {
-                f.compiledFile = f.parsed
```

Comment:
This was one of the biggest issues - imagine user paused typing, the cp.compiler.compileFiles is called, and while it is running, the user types some more. Then (before cp.compiler.compileFiles finishes), the client calls e.g. documentSymbols, which will set f.parsed to the newest version of the document's content. Then cp.compiler.compileFiles finishes, and this code here will set compiledFile to the newest version (so that subsequent lintting won't recompile the document), but also persists the old context and container - which will then be used to send diagnostics.